### PR TITLE
[ConvDiff] Laplacian Embedded Element

### PR DESCRIPTION
**Pull request by frawahl**
**Description**
Added a Laplacian embedded element for the `ConvectionDiffusionApplication`. It uses the same modified shape functions and element splitting for intersected elements as the embedded element of the `FluidDynamicsApplication`. For an intersected element, the LHS and RHS are calculated by computing a volume integral  on the positive side of the element, an interface integral at the intersection boundary, as well as Nitsche terms for the weak imposition of a Dirichlet boundary condition.

**Changelog**
- Added laplacian embedded element to the `ConvectionDiffusionApplication`
- Added variable `PENALTY_DIRICHLET` to the `ConvectionDiffusionApplication`
- Added variable `EMBEDDED_SCALAR` to the `ConvectionDiffusionApplication` for the Dirichlet boundary value in intersected elements
- Added a cpp test of the element to the `ConvectionDiffusionApplication`

### CODE REVIEW

File: kratos/includes/variables.h [hunk 1]
```diff
@@ -126,6 +126,8 @@ namespace Kratos
 
     KRATOS_DEFINE_VARIABLE( double, TIME_INTEGRATION_THETA )
```
Comment (philbucher) [inline kratos/includes/variables.h]: Can you please put this to the ConDiffApp if it is only used there?

Comment (frawahl) [reply kratos/includes/variables.h]: Yes, for now it is, I moved it.

File: applications/ConvectionDiffusionApplication/CMakeLists.txt [hunk 2]
```diff
@@ -14,6 +14,7 @@ set( KRATOS_CONVECTION_DIFFUSION_APPLICATION_CORE_SOURCES
     ${CMAKE_CURRENT_SOURCE_DIR}/custom_elements/conv_diff_2d.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/custom_elements/conv_diff_3d.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/custom_elements/laplacian_element.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/custom_elements/laplacian_embedded_element.cpp
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/CMakeLists.txt]: I'd call the files `embedded_laplacian_element.cpp` and `embedded_laplacian_element.h` to follow the convention we stated in the `FluidDynamicsApplication`. Same applies to the class names.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 3]
```diff
@@ -0,0 +1,433 @@
+// KRATOS ___ ___  _  ___   __   ___ ___ ___ ___
+//       / __/ _ \| \| \ \ / /__|   \_ _| __| __|
+//      | (_| (_) | .` |\ V /___| |) | || _|| _|
+//       \___\___/|_|\_| \_/    |___/___|_| |_|  APPLICATION
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ruben Zorrilla
+//                   Franziska Wahl
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]:
```suggestion
//  Main authors:    Franziska Wahl
//
```
I think you did everything in here.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/checks.h"
+#include "includes/define.h"
+#include "includes/kratos_flags.h"
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]:
```suggestion
```
I think this shouldn't be required.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+#include "includes/convection_diffusion_settings.h"
+#include "utilities/geometry_utilities.h"
+#include "utilities/element_size_calculator.h"
+
+// Application includes
+#include "custom_elements/laplacian_embedded_element.h"
+#include "convection_diffusion_application_variables.h"
+
+#include "modified_shape_functions/triangle_2d_3_modified_shape_functions.h"
+#include "modified_shape_functions/tetrahedra_3d_4_modified_shape_functions.h"
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]: I'd move these to the `// Project includes` section above.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/checks.h"
+#include "includes/define.h"
+#include "includes/kratos_flags.h"
+#include "includes/ublas_interface.h"
+#include "includes/variables.h"
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]:
```suggestion
```
I think these two are not needed (and are pretty heavy headers).

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+
+    // Check if the element belongs to the intersected ones
+    if (data.IsSplit()) {
+
+        // Get nodal distances and set splitting as well as shape functions
+        InitializeGeometryData(data);
+
+        // Resizing and resetting the LHS
+        if(rLeftHandSideMatrix.size1() != NumNodes)
+            rLeftHandSideMatrix.resize(NumNodes,NumNodes,false);
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]:
```suggestion
        if (rLeftHandSideMatrix.size1() != NumNodes || rLeftHandSideMatrix.size2() != NumNodes) {
            rLeftHandSideMatrix.resize(NumNodes,NumNodes,false);
        }
```

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+    VectorType& rRightHandSideVector,
+    const ProcessInfo& rCurrentProcessInfo,
+    const EmbeddedElementData& rData)
+{
+    const auto& r_geom = GetGeometry();
+    auto& r_settings = *rCurrentProcessInfo[CONVECTION_DIFFUSION_SETTINGS];
+
+    const Variable<double>& r_unknown_var = r_settings.GetUnknownVariable();
+    const Variable<double>& r_diffusivity_var = r_settings.GetDiffusionVariable();
+    const Variable<double>& r_volume_source_var = r_settings.GetVolumeSourceVariable();
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]:
```suggestion
    const auto& r_unknown_var = r_settings.GetUnknownVariable();
    const auto& r_diffusivity_var = r_settings.GetDiffusionVariable();
    const auto& r_volume_source_var = r_settings.GetVolumeSourceVariable();
```

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+    MatrixType& rLeftHandSideMatrix,
+    VectorType& rRightHandSideVector,
+    const ProcessInfo& rCurrentProcessInfo,
+    const EmbeddedElementData& rData)
+{
+    const auto& r_geom = GetGeometry();
+    auto& r_settings = *rCurrentProcessInfo[CONVECTION_DIFFUSION_SETTINGS];
+
+    const Variable<double>& r_unknown_var = r_settings.GetUnknownVariable();
+    const Variable<double>& r_diffusivity_var = r_settings.GetDiffusionVariable();
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]:
```suggestion
    const auto& r_unknown_var = r_settings.GetUnknownVariable();
    const auto& r_diffusivity_var = r_settings.GetDiffusionVariable();
```

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]:
```suggestion
    const auto& r_unknown_var = r_settings.GetUnknownVariable();
    const auto& r_diffusivity_var = r_settings.GetDiffusionVariable();
```

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+        rData.PositiveInterfaceWeights,
+        this->GetIntegrationMethod());
+
+    // Positive side interface normals
+    p_calculator->ComputePositiveSideInterfaceAreaNormals(
+        rData.PositiveInterfaceUnitNormals,
+        this->GetIntegrationMethod());
+
+    // Normalize the interface normals
+    const double h = ElementSizeCalculator<TTDim,NumNodes>::MinimumElementSize(this->GetGeometry());
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]: Why the `MinimumElementSize`? I'd use the `AverageElementSize` as default.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+        nodal_conductivity[n] = r_geom[n].FastGetSolutionStepValue(r_diffusivity_var);
+        temp[n] = r_geom[n].GetSolutionStepValue(r_unknown_var);
+    }
+
+    // Nitsche penalty constant
+    const double gamma = rCurrentProcessInfo[PENALTY_DIRICHLET]; 
+    // Dirichlet boundary value
+    const double temp_bc = GetValue(EMBEDDED_SCALAR);
+    // Measure of element size
+    const double h = ElementSizeCalculator<TTDim,NumNodes>::MinimumElementSize(r_geom);
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]: Same comment. Why the `MinimumElementSize`?

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h [hunk 13]
```diff
@@ -0,0 +1,390 @@
+// KRATOS ___ ___  _  ___   __   ___ ___ ___ ___
+//       / __/ _ \| \| \ \ / /__|   \_ _| __| __|
+//      | (_| (_) | .` |\ V /___| |) | || _|| _|
+//       \___\___/|_|\_| \_/    |___/___|_| |_|  APPLICATION
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ruben Zorrilla
+//                   Franziska Wahl
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h]:
```suggestion
//  Main authors:    Franziska Wahl
//
```

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+// Forward declaration of data container class
+namespace LaplacianEmbeddedInternals {
+    template<std::size_t TDim> class EmbeddedElementData;
+}
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h]: I think you can avoid this forward declaration by moving the `EmbeddedElementData` declaration from the bottom part of the file to the top.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+        NodalDistances.resize(NumNodes);
+    }
+    for (std::size_t i = 0; i < NumNodes; ++i) {
+        NodalDistances[i] = r_geom[i].FastGetSolutionStepValue(DISTANCE);
+    }
+
+    // Number and indices of positive and negative distance function values
+    NumPositiveNodes = 0;
+    NumNegativeNodes = 0;
+    PositiveIndices.clear();
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]:
```suggestion
```
I think positive indices are not needed in this case.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp [hunk 16]
```diff
@@ -0,0 +1,433 @@
...
+
+    for (std::size_t i = 0; i < NumNodes; ++i){
+        if (NodalDistances[i] > 0.0){
+            NumPositiveNodes++;
+            PositiveIndices.push_back(i);
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]:
```suggestion
```
I think positive indices are not needed in this case.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+    Matrix PositiveSideN;
+    ShapeFunctionsGradientsType PositiveSideDNDX;
+    Vector PositiveSideWeights;
+
+    Matrix PositiveInterfaceN;
+    ShapeFunctionsGradientsType PositiveInterfaceDNDX;
+    Vector PositiveInterfaceWeights;
+    InterfaceNormalsType PositiveInterfaceUnitNormals;
+
+    std::vector< size_t > PositiveIndices;
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h]:
```suggestion
```
I think positive indices are not needed in this case.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+    ///@name Public Operations
+    ///@{
+
+    /**
+     * @brief Split element data container initialization
+     * This method initializes the embedded formulation data container. This implies to get the nodal distances.
+     */
+    void Initialize(
+        const Element& rElement
+    );
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h]:
```suggestion
    void Initialize(const Element& rElement);
```
One line for single argument functions (style guide).

File: applications/ConvectionDiffusionApplication/tests/cpp_tests/test_laplacian_embedded_element.cpp [hunk 19]
```diff
@@ -0,0 +1,123 @@
+// KRATOS ___ ___  _  ___   __   ___ ___ ___ ___
+//       / __/ _ \| \| \ \ / /__|   \_ _| __| __|
+//      | (_| (_) | .` |\ V /___| |) | || _|| _|
+//       \___\___/|_|\_| \_/    |___/___|_| |_|  APPLICATION
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ruben Zorrilla
+//                   Franziska Wahl
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/tests/cpp_tests/test_laplacian_embedded_element.cpp]:
```suggestion
//  Main authors:    Franziska Wahl
//
```

File: applications/ConvectionDiffusionApplication/tests/cpp_tests/test_laplacian_embedded_element.cpp [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+#include "convection_diffusion_application_variables.h"
+
+
+namespace Kratos {
+namespace Testing
+{
+    void SetLaplacianEmbeddedElementTestModelPart(ModelPart &rModelPart)
+    {
+        // Set buffer size
+        rModelPart.SetBufferSize(2);
```
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/tests/cpp_tests/test_laplacian_embedded_element.cpp]:
```suggestion
        rModelPart.SetBufferSize(1);
```
I think that 1 must be enough in this case.

**Review by rubenzorrilla — COMMENTED** [review]
Only a very few minor comments. However, I think that we can eventually simplify the design by defining the data container inside the element class (future work).

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h (hunk 14 above)
Comment (frawahl) [reply applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.h]: True, I just thought it's nicer the other way round.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp (hunk 11 above)
Comment (frawahl) [reply applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]: Because it's consistent with the embedded fluid element.

File: applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp (hunk 12 above)
Comment (frawahl) [reply applications/ConvectionDiffusionApplication/custom_elements/laplacian_embedded_element.cpp]: Also here the MinimumElementSize is used in the embedded fluid element..

**Review by rubenzorrilla — APPROVED** [review]
